### DISCUSSION TITLE
Improved support for writing new algorithms in application code

### DISCUSSION THREAD

**Discussion by leshabirukov** [HeWeMel/nographs#21] [Proposed features that need discussions]
>> Suggestion: Computing strongly connected components (of the sub-graph reachable from a set of vertices), e.g., with algorithms like described [here](https://pure.mpg.de/rest/items/item_2344001_7/component/file_2456700/content). These are linear-time algorithms that can be applied to directed graphs.

_Originally posted by @HeWeMel in https://github.com/HeWeMel/nographs/discussions/12#discussioncomment-7514071_

Hello again!  
New issue, and now it's real deal, I think. Consider Tarjan algorithm: With DFS, we need to do some work on_leaving node, but how? The only visitor is 'next_something( vert )', and we already leaved it!  
So, we in need of 'on_leave(v)'. What we can do (in user code part) now, is everytime leaving vertex, check it's predecessors, possibly all way to root. Not the dream.  
Possible solutions:  
- Add algorithms like Tarjans to the nog core. That's what you actuallu proposed, but I think it must be user level.
- Saying "we don't". Painfuly restrictive.
- Point to solution, using interaction with different graph lib. Hardly could be done with good performance.
- Provide traversal object with 'on_leave' (on_finish, post_process) event, user can hook.
- Provide some other template for algoritms like Tarjan's.

**Comment by HeWeMel** [comment]
Approaches  4 & 5 are possible (maybe in addition to 1).

**Reply by HeWeMel** [reply]
(A Tarjan implemented as part of the library can offer full typing flexibility and good performance, better than what can be done in  application code)

**Reply by leshabirukov** [reply]
> ...as part of the library can offer...  

Of course it is, the question is lib structure complexity cost. In my opinion, general solution with acceptable performance is better then specialized with the best performance, at least for begining.

**Reply by HeWeMel** [reply]
Most libraries do both. Thus, I evaluate both.

**Comment by HeWeMel** [comment]
Some libraries provide a DFS that several visitors can be given to. Additional chance: Their return values can *control* how the DFS continues. Disadvantage: In Python, function calls are quite slow. But since the application can decide whether to use this functionality or not, this might not be a problem.

**Reply by HeWeMel** [reply]
Side note: It could be, that all the necessary functionality is already given in the implementation of the topological search, as it has some job to do when a vertex is *left*.

**Comment by HeWeMel** [comment]
A second approach could be to report vertices not just when they have a single, fixed state (e.g., *visited*), but in several states, together with the state. In order to keep the interface and typing constant, the state could be given as attribute of the traversal. This requires an attribute write (in the lib) and an attribute read (in the application) instead of a function call.

**Comment by HeWeMel** [comment]
Both approaches could either be done by adding further parameters to existing functions (callback function or flags demanding additional functionality) or by offering an additional traversal algorithm with the additional features.

I do not know so far, which approach in which form is the best for the users...

**Reply by leshabirukov** [reply]
> additional traversal algorithm  

There is already a few of DFS traversals in the `nog`, and, I suppose any of them could benefit from SCC, aaand I sence a ghost of the combinatorial growth.

**Reply by leshabirukov** [reply]
> which form is the best for the users...

“Put the oxygen mask on yourself first and then on the baby.”  
User will not benefit if you sank with the support.

**Reply by HeWeMel** [reply]
The algorithms of NoGraphs are non-recursive. In this variant, there is no notion of "leaving" a vertex, if not implemented manually.  See, e.g., [simple version of non-recursive DFS in wikipia](https://en.wikipedia.org/wiki/Depth-first_search#:~:text=procedure%20DFS_iterative(G%2C%20v)%20is%0A%20%20%20%20let%20S%20be%20a%20stack%0A%20%20%20%20S.push(v)%0A%20%20%20%20while%20S%20is%20not%20empty%20do%0A%20%20%20%20%20%20%20%20v%20%3D%20S.pop()%0A%20%20%20%20%20%20%20%20if%20v%20is%20not%20labeled%20as%20discovered%20then%0A%20%20%20%20%20%20%20%20%20%20%20%20label%20v%20as%20discovered%0A%20%20%20%20%20%20%20%20%20%20%20%20for%20all%20edges%20from%20v%20to%20w%20in%20G.adjacentEdges(v)%20do%20%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20S.push(w))

So, implementing the feature costs code, runtime, and memory. Either in form of a new traversal class, or in form of a new algorithm within a class, or a conditionally activated additional functionality (which means more complexity within an algorithm).

**Reply by HeWeMel** [reply]
Example: [networx ](https://networkx.org/documentation/stable/_modules/networkx/algorithms/traversal/depth_first_search.html#dfs_preorder_nodes:~:text=if%20depth_now%20%3C%20depth_limit%3A-,stack.append((child%2C%20iter(G%5Bchild%5D))),-depth_now%20%2B%3D%201%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20break%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20else) pushes tuples of (vertex, iterator_of_children) on the stack, instead of just the vertex. Tuples are large, iterators also. And the lib gets each such tuple several times from the stack in order to process more children. This is quite ineffective. But by doing so, a good bases is there to derive [all kinds of special functions](https://networkx.org/documentation/stable/reference/algorithms/traversal.html).
